google-cloud-spanner/src/test/java/com/google/cloud/spanner/InstanceConfigTest.java

```diff
@@ -63,26 +76,44 @@ public void testInstanceConfigFromProto() {
 
     assertEquals(
         new InstanceConfig(
-            InstanceConfigId.of("my-project", "my-instance-config"),
-            "Display Name",
-            Arrays.asList(
-                ReplicaInfo.newBuilder()
-                    .setLocation("Replica Location 1")
-                    .setType(ReplicaInfo.ReplicaType.READ_WRITE)
-                    .setDefaultLeaderLocation(true)
-                    .build(),
-                ReplicaInfo.newBuilder()
-                    .setLocation("Replica Location 2")
-                    .setType(ReplicaInfo.ReplicaType.READ_ONLY)
-                    .setDefaultLeaderLocation(false)
-                    .build(),
-                ReplicaInfo.newBuilder()
-                    .setLocation("Replica Location 3")
-                    .setType(ReplicaInfo.ReplicaType.WITNESS)
-                    .setDefaultLeaderLocation(false)
-                    .build()),
-            Arrays.asList("Leader Option 1", "Leader Option 2"),
-            client),
+            (InstanceConfig.Builder)
```

Comment:
I don't understand why we need a cast here?

Reply:
All the setters exist in base InstanceConfigInfo.Builder/BuilderImpl class and return InstanceConfigInfo.Builder object. We just inherit these setters in InstanceConfig.Builder. When we use these setters, the object returned is of type InstanceConfigInfo.Builder and we cast it to more specific InstanceConfig.Builder.

Reply:
Ok, that makes sense, but I think that shows a small (potential?) problem with the current implementation: Users should not be forced to do any casts when working with the builder. So if the implementation in the abstract base class (`InstanceConfigInfo`) returns the abstract class, but all our public methods require the use of the concrete subclass (`InstanceConfig`), then we should make sure that the methods that are visible to users als return an instance of `InstanceConfig.Builder` and not `InstanceConfigInfo.Builder`.
Could you take a look at `InstanceConfig` and `InstanceConfigInfo` how that is handled there, and modify this test case (and potentially the implementation) to show that users do not need to do any casting when working with the builder(s)?

Reply:
Thanks for the explanation! 

The solution I've implemented now to handle this is to override the setters in InstanceConfig.Builder class and make them return InstanceConfig.Builder.

For example, overridden setDisplayName in InstanceConfig.Builder would look something like this:
```
@Override
public Builder setDisplayName(String displayName) {
  super.setDisplayName(displayName);
  return this;
}
```